drivers: hwinfo: native: report reset cause

tpambor opened the pull request:
This PR adds support for reporting the reset cause to native_sim.

The default is to report POR (Power-On Reset).
If the system is rebooted using `sys_reboot()`, the reset cause is set to SOFTWARE.

aescolar committed:
board native_sim: Test sys_reboot and hw_info reporting the reset cause

Add a test of the CONFIG_NATIVE_SIM_REBOOT functionality together
with the hw_info get_reset_cause() logic.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar reviewed (changes requested):
Question out of curiosity, are you doing this change because you have a need for it? or just because you thought it may be nice for somebody to have it?

Requested changes:
* Let's also add a test for this (I just added one here, you are more than welcome to take it):
https://github.com/aescolar/zephyr/commit/688ca66f2f945f61b448fe56bc485e487a5875f2

* Note a few requests for changes below.

And a general comment which does not need any action, but just a remark:
_Right now, this HW_info peripheral can be built for any posix arch based target, but the reboot logic is optional and only for the native_sim target.
This means this hw_info reset cause code will only do what is expected in the native_sim case when CONFIG_NATIVE_SIM_REBOOT is enabled.
Note, in tree, we only have the native_sim and nrf*bsim targets. And the nrf*bsim targets are likely to have the real nrf hw info driver enabled at some point.
This is not a problem beyond being potentially confusing for users._

aescolar commented:
CC @koalatux

tpambor commented on scripts/native_simulator/common/src/include/nsi_host_trampolines.h:
```diff
@@ -37,6 +37,8 @@ void *nsi_host_realloc(void *ptr, unsigned long size);
 void nsi_host_srandom(unsigned int seed);
 char *nsi_host_strdup(const char *s);
 long nsi_host_write(int fd, const void *buffer, unsigned long size);
+char *nsi_host_getenv(const char *name);
```
Sure, I opened https://github.com/BabbleSim/native_simulator/pull/8

tpambor committed:
native_simulator: Get latest from upstream

Align with native_simulator's upstream main
4eab13716376e63236d77013f996a897d24dd780

Which includes:
4eab137 Host trampolines: Add getenv/setenv

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor committed:
drivers: hwinfo: native: report reset cause

Support reporting the reset cause for native_sim. The default is to
report POR (Power-On Reset). If CONFIG_NATIVE_SIM_REBOOT was enabled and
the system is rebooted using sys_reboot(), the reset cause is set to
SOFTWARE.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

alexanderwachter reviewed (commented):
Nice to have hwinfo in native_sim!
Thanks!
